cmd/checkpoint/main.go

```diff
@@ -19,234 +21,499 @@ import (
+			// This means the kubelet thinks it should still be running, which has the same scheduling info that we do --
+			// so we won't make any decisions about its checkpoint.
+			if _, ok := localParentPods[id]; ok {
+				glog.V(4).Infof("API GC: skipping inactive checkpoint %s", id)
+				continue
 			}
-		case kubeSystemAPIServerRunning(podList, client):
-			glog.Info("kube-apiserver found, creating temp-apiserver manifest")
-			// The self-hosted API Server is running. Let's snapshot the pod,
-			// clean it up a bit, and then save it to the ignore path for
-			// later use.
-			tempAPIServerManifest.Spec = parseAPIPodSpec(podList)
-			convertSecretsToVolumeMounts(client, &tempAPIServerManifest)
-			writeManifest(tempAPIServerManifest)
-			glog.Infof("finished creating temp-apiserver manifest at %s\n", checkpointManifest)
-
-		default:
-			glog.Info("no apiserver running, installing temp apiserver static manifest")
-			b, err := ioutil.ReadFile(checkpointManifest)
-			if err != nil {
-				glog.Error(err)
-			} else {
-				if err := ioutil.WriteFile(activeManifest, b, 0644); err != nil {
-					glog.Error(err)
-				}
+
+			// If the inactive checkpoint does not have a parent in the api-server, we must assume it should no longer be running on this node.
+			// NOTE: It's possible that a replacement for this pod has not been rescheduled elsewhere, but that's not something we can base our decision on.
+			//       For example, if a single scheduler is running, and the node is drained, the scheduler pod will be deleted and there will be no replacement.
+			//       However, we don't know this, and as far as the checkpointer is concerned - that pod is no longer scheduled to this node.
+			if _, ok := apiParentPods[id]; !ok {
+				glog.V(4).Infof("API GC: should remove inactive checkpoint %s", id)
+				removeMap[id] = struct{}{}
+				delete(inactiveCheckpoints, id)
+			}
+		}
+
+		// Scan active checkpoints we should GC
+		for id := range activeCheckpoints {
+			// If the active checkpoint does not have a parent in the api-server, we must assume it should no longer be running on this node.
+			if _, ok := apiParentPods[id]; !ok {
+				glog.V(4).Infof("API GC: should remove active checkpoint %s", id)
+				removeMap[id] = struct{}{}
+				delete(activeCheckpoints, id)
 			}
 		}
-		time.Sleep(1 * time.Second)
 	}
+
+	// Can make decisions about starting/stopping checkpoints just with local state.
+
+	// If there is an inactive checkpoint, and no parent is running, start the checkpoint
+	for id := range inactiveCheckpoints {
+		if _, ok := localParentPods[id]; !ok {
+			glog.V(4).Infof("Should start checkpoint %s", id)
+			start = append(start, id)
+		}
+	}
+
+	// If there is a parent pod and an active checkpoint, stop the active checkpoint
+	// The parent may not be in a running state, but the kubelet is trying to start it so we should get out of the way.
+	for id := range localParentPods {
+		if _, ok := activeCheckpoints[id]; ok {
```

Comment:
range over the activeCheckpoints in the outer loop? be consistent with line 148?